### ISSUE TITLE
ServiceTemplate Lifecycle Management (goals, desired UX specification, etc.)

### ISSUE THREAD

**Issue by DinaBelova** [Mirantis/hmc#248]
**Goals**

In #203  we have implemented state management inside a cluster Template, while this solves some of the use cases of a platform engineering teams and is quite simple to use, it has one big downside: there is no centrals place to define a `helmchart/artifact` which should be deployed across all clusters, no matter which template they use. Instead you need to define the `helmchart/artifact` in each template, and if you have many templates updating a single helm chart across all these templates becomes a pain very fast.

The goal of this epic is to allow the platform leads to define, control, and manage the lifecycle of (beachhead) service-specific templates, such as Ingress controllers or other application-level components, while allowing platform engineers to implement and manage those templates across environments.

Goals:
- **_Separation of Concerns_**: Decouple the management of service/beachhead templates (e.g., Ingress, monitoring solutions) from cluster templates.
- **_ServiceTemplate Lifecycle Management_**: Allow templates for services to be version-controlled, updated, and managed similarly to ClusterTemplates.
- **_Access Control_**: Provide granular control for platform leads over who can use which templates, ensuring service templates are only available to approved environments.
- **_Upgrade Path_**: Allow platform engineers to selectively upgrade service templates, providing a clear upgrade path from one template version to the next.

**Major deliverables**
- **_ServiceTemplate and ServiceTemplateChain Objects_**:
    - Define Helm chart-based templates for services (e.g., beachhead services) using ServiceTemplate objects.
    - Define ServiceTemplateChain objects to manage the supported versions of service templates and their upgrade paths.
- **_Versioning and Immutability_**:
    - ServiceTemplates must be immutable. Each new version of a service must create a new ServiceTemplate object, similar to the ClusterTemplate system.
    - Platform leads must define which versions are available and manage the upgrade path.
- **_Namespace-Based Access Control_**:
    - Implement a system for controlling which namespaces (environments like dev, test, or prod) can access specific versions of ServiceTemplates.
    - Granular access control should ensure that services are only deployed in approved environments.
- **_Upgrade Mechanism_**:
    - Platform engineers must have the ability to choose when to apply a service template upgrade, ensuring there is no unintended impact on running clusters.
- **_TemplateAccess Management_**:
    - Define and implement the TemplateManagement object that governs access to both ClusterTemplateChains and ServiceTemplateChains across namespaces. 

**Who it benefits**
- **_Platform Leads_**: Can control the availability and lifecycle of (beachhead) service templates, ensuring that engineers use only approved and tested service versions.
- **_Platform Engineers_**: Gain the ability to manage services like Ingress, logging, or monitoring templates independently of cluster upgrades, allowing for more flexible management of the overall stack.

**Acceptance criteria**

**_New ServiceTemplate_**:
- Platform leads can create new ServiceTemplate objects and associate them with ServiceTemplateChains.
- New service templates must be created when an updated Helm chart version is available (e.g., mirantis-ha-ingress-nginx v1.11.2).

**_ServiceTemplateChain_**:
- Platform leads can define and update ServiceTemplateChains that specify which versions are supported and what upgrades are available for each template.
- Available upgrades for a ServiceTemplate must be clearly defined, allowing platform engineers to see what options they have for upgrading.

**_ServiceTemplate Updates_**:
- Platform engineers must be able to see when a new service template version is available and apply upgrades to the services independently of the cluster.
- TBD:  Updates must be non-disruptive, with the option for engineers to approve upgrades before applying them.

**_Template Access_**:
- Platform leads can assign ServiceTemplates and ServiceTemplateChains to specific namespaces, ensuring that only approved environments (e.g., dev, test, prod) have access to the correct templates.
- New ServiceTemplates created by platform leads or Mirantis are inaccessible to platform engineers until explicitly approved and made available by the platform lead.

**Assumptions**

- ServiceTemplates (Helm charts and CRDs) are immutable, and a new template must be created for every new version.
- TBD multiple template repositories (e.g., Mirantis and customer-specific) will be supported.

_Note:  The actual deployment and updates of services will be managed by the [hmc.mirantis.com](https://github.com/Mirantis/hmc) system._

**Limitations**

The epic focuses solely on managing the lifecycle of service templates and the associated access control. It does not cover the validation of the deployed services in the cluster or the monitoring of their health post-deployment.

**Out of Scope**
- Validation of the deployed services.
- The actual  BHS Helm charts to be implementation or development of service applications.
- Cluster-level lifecycle management (covered in #296 )

**User stories**
- Platform Lead: Define ServiceTemplateChain
    - As a platform lead, I want to be able to create and manage ServiceTemplateChains for different services (e.g., Ingress controllers) so that I can define supported versions and available upgrades.
- Platform Lead: Manage Service Template Access
    - As a platform lead, I need to control which ServiceTemplates are accessible to specific namespaces (e.g., dev, test, prod) to ensure only approved templates are used.
- Platform Engineer: Upgrade ServiceTemplates
    - As a platform engineer, I want to upgrade the services (e.g., BHS) in my cluster using the latest approved templates so that I can keep my services up to date.
- Platform Engineer: View ServiceTemplate Upgrades
    - As a platform engineer, I need to see which service template upgrades are available so that I can plan and apply updates when needed.
- Platform Engineer: Non-disruptive Upgrades
    - As a platform engineer, I want to apply service template upgrades in a non-disruptive way, ensuring that my running clusters and services are not affected unless approved.

Related implementation GH issues:
- #268
- #316  (depends on #266)

**Comment by DinaBelova** [comment]
all actual implementation tasks (for first level of development) completed -> closing this epic where we tracked the original goals / concepts / etc.
